File: src/main/java/dev/klomptech/jbark/config/NameHints.java
package dev.klomptech.jbark.config;

import java.util.Collection;

// Hints for dog-themed error messages
public final class NameHints {

  private static final int MAX_DISTANCE = 2;

  private NameHints() {}

  public static String suggest(final String name, final Collection<String> candidates) {
    if (name == null || candidates == null || candidates.isEmpty()) {
      return null;
    }
    String normalized = ConfigLoader.normalise(name);
    String phrase = ConfigLoader.resolveNamePhrase(normalized);
    if (phrase != null) {
      normalized = phrase;
    }
    String best = null;
    int bestDistance = MAX_DISTANCE + 1;
    for (String candidate : candidates) {
      int distance = distance(normalized, candidate);
      if (distance < bestDistance) {
        bestDistance = distance;
        best = candidate;
      }
    }
    return bestDistance <= MAX_DISTANCE ? best : null;
  }

  public static String hintPhrase(final String name, final Collection<String> candidates) {
    String suggestion = suggest(name, candidates);
    if (suggestion == null || suggestion.equals(ConfigLoader.normalise(name))) {
      return "";
    }
    String display = suggestion.contains("_") ? suggestion.replace('_', ' ') : suggestion;
    return " Did you mean " + display + "?";
  }

  private static int distance(final String left, final String right) {
    int[][] table = new int[left.length() + 1][right.length() + 1];
    for (int i = 0; i <= left.length(); i++) {
      table[i][0] = i;
    }
    for (int j = 0; j <= right.length(); j++) {
      table[0][j] = j;
    }
    for (int i = 1; i <= left.length(); i++) {
      for (int j = 1; j <= right.length(); j++) {
        int cost = left.charAt(i - 1) == right.charAt(j - 1) ? 0 : 1;
        table[i][j] =
            Math.min(
                Math.min(table[i - 1][j] + 1, table[i][j - 1] + 1), table[i - 1][j - 1] + cost);
      }
    }
    return table[left.length()][right.length()];
  }
}

File: pbark/pbark/config/name_hints.py
from __future__ import annotations

import json
from pbark.config.config_loader import ConfigLoader


class NameHints:
    MAX_DISTANCE = 2

    @staticmethod
    def suggest(name: str | None, candidates) -> str | None:
        if name is None or candidates is None or not candidates:
            return None
        normalized = ConfigLoader.normalise(name)
        phrase = ConfigLoader.resolve_name_phrase(normalized)
        if phrase is not None:
            normalized = phrase
        best = None
        best_distance = NameHints.MAX_DISTANCE + 1
        for candidate in candidates:
            distance = NameHints._distance(normalized, candidate)
            if distance < best_distance:
                best_distance = distance
                best = candidate
        return best if best_distance <= NameHints.MAX_DISTANCE else None

    @staticmethod
    def hint_phrase(name: str, candidates) -> str:
        suggestion = NameHints.suggest(name, candidates)
        if suggestion is None or suggestion == ConfigLoader.normalise(name):
            return ""
        display = suggestion.replace("_", " ") if "_" in suggestion else suggestion
        return f" Did you mean {display}?"

    @staticmethod
    def _distance(left: str, right: str) -> int:
        table = [[0] * (len(right) + 1) for _ in range(len(left) + 1)]
        for i in range(len(left) + 1):
            table[i][0] = i
        for j in range(len(right) + 1):
            table[0][j] = j
        for i in range(1, len(left) + 1):
            for j in range(1, len(right) + 1):
                cost = 0 if left[i - 1] == right[j - 1] else 1
                table[i][j] = min(
                    table[i - 1][j] + 1,
                    table[i][j - 1] + 1,
                    table[i - 1][j - 1] + cost,
                )
        return table[len(left)][len(right)]
